File: lib/model/music_info.dart

import 'dart:typed_data';

class MusicInfo {
  final String title;
  final String artist;
  final Uint8List? albumArt;
  final bool isPlay;

  const MusicInfo({
    required this.title,
    required this.artist,
    this.albumArt,
    this.isPlay = false,
  });

  MusicInfo.fromMap(Map<String, dynamic> map)
      : title = map['title'] ?? "",
        artist = map['artist'] ?? "",
        isPlay = map['isPlay'] == 1 ? true : false,
        albumArt = map['albumArt'];

  static const MusicInfo empty = MusicInfo(
    title: '',
    artist: '',
    isPlay: false,
  );

  @override
  String toString() {
    return 'title: $title, artist: $artist, album: ${albumArt?.length}, isPlay: $isPlay';
  }
}

File: android/app/src/main/kotlin/com/ohmnia/car_music_info/model/MusicInfo.kt
package com.ohmnia.car_music_info.model

import android.graphics.Bitmap
import java.io.ByteArrayOutputStream

data class MusicInfo(
    val title: String,
    val artist: String,
    val albumArt: Bitmap?,
    val isPlay: Boolean,
) {
    companion object {
        val empty = MusicInfo("", "", null, false)

        fun byteFromBitmap(albumArt: Bitmap?): ByteArray? {
            return albumArt?.let { art ->
                val stream = ByteArrayOutputStream()
                art.compress(Bitmap.CompressFormat.PNG, 100, stream)
                stream.toByteArray()
            }
        }
    }

    fun toMap(): Map<String, Any?> {
        return mutableMapOf<String, Any?>().apply {
            put("title", title)
            put("artist", artist)
            put("albumArt", byteFromBitmap(albumArt))
            put("isPlay", if (isPlay) 1 else 0)
        }
    }

    override fun equals(other: Any?): Boolean {
        return other is MusicInfo
                && other.title == title
                && other.artist == artist
                && other.isPlay == isPlay
                && other.getBitmapSize() == getBitmapSize()
    }

    private fun getBitmapSize() = albumArt?.byteCount ?: 0

    override fun hashCode(): Int {
        var result = title.hashCode()
        result = 31 * result + artist.hashCode()
        result = 31 * result + getBitmapSize()
        result = 31 * result + isPlay.hashCode()
        return result
    }
}
